lib/handlers/allow.js

```diff
@@ -12,59 +15,65 @@ process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
 
 function allow (mode) {
   return function allowHandler (req, res, next) {
-    var ldp = req.app.locals.ldp
+    let ldp = req.app.locals.ldp
     if (!ldp.webid) {
       return next()
     }
-    var baseUri = utils.uriBase(req)
+    let baseUri = utils.uriBase(req)
+
+    let fetchDocument = documentFetcher(req.hostname, ldp, baseUri)
 
     var acl = new ACL({
       debug: debug,
-      fetch: fetchDocument(req.hostname, ldp, baseUri),
+      fetch: fetchDocument,
       suffix: ldp.suffixAcl,
       strictOrigin: ldp.strictOrigin
     })
 
-    getUserId(req, function (err, userId) {
-      if (err) return next(err)
-
-      var reqPath = res && res.locals && res.locals.path
-        ? res.locals.path
-        : req.path
-      ldp.exists(req.hostname, reqPath, (err, ret) => {
-        if (ret) {
-          var stat = ret.stream
-        }
-        if (!reqPath.endsWith('/') && !err && stat.isDirectory()) {
-          reqPath += '/'
-        }
-        var options = {
-          origin: req.get('origin'),
-          host: req.protocol + '://' + req.get('host')
-        }
-        return acl.can(userId, mode, baseUri + reqPath, next, options)
+    let webId = req.session.userId
+    let onBehalfOf = utils.debrack(req.get('On-Behalf-Of'))
+    getUserId(webId, onBehalfOf, fetchDocument)
+      .then(userId => {
+        let reqPath = res && res.locals && res.locals.path
+          ? res.locals.path
+          : req.path
+        ldp.exists(req.hostname, reqPath, (err, ret) => {
+          if (ret) {
+            var stat = ret.stream
+          }
+          if (!reqPath.endsWith('/') && !err && stat.isDirectory()) {
+            reqPath += '/'
+          }
+          var options = {
+            origin: req.get('origin'),
+            host: req.protocol + '://' + req.get('host')
+          }
+          return acl.can(userId, mode, baseUri + reqPath, next, options)
+        })
+      })
+      .catch(err => {
+        next(err)
```

Comment:
I'm not too familiar with express. Does this pass the `err` object to the next middleware?